### PR TITLE
Allow rolling aggregations for window functions

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
It is WIP because I haven't run any tests.

I did simple stupid benchmark on a single node Presto server. Results are below, second run was executed on Presto without this patch:
```
presto:sf1> select count(a) from (SELECT sum(quantity) OVER (ROWS BETWEEN 2000 PRECEDING AND 2000 FOLLOWING) from lineitem) t(a);
  _col0
---------
 6001215
(1 row)

Query 20190312_185636_00020_9axmd, FINISHED, 1 node
Splits: 38 total, 38 done (100.00%)
0:02 [6M rows, 0B] [2.53M rows/s, 0B/s]

presto:sf1> select count(a) from (SELECT sum(quantity) OVER (ROWS BETWEEN 2000 PRECEDING AND 2000 FOLLOWING) from lineitem) t(a);
Query 20190312_185752_00000_ysa4v failed: No nodes available to run query

presto:sf1> select count(a) from (SELECT sum(quantity) OVER (ROWS BETWEEN 2000 PRECEDING AND 2000 FOLLOWING) from lineitem) t(a);
  _col0
---------
 6001215
(1 row)

Query 20190312_185756_00001_ysa4v, FINISHED, 1 node
Splits: 38 total, 38 done (100.00%)
5:03 [6M rows, 0B] [19.8K rows/s, 0B/s]
```

**Comment by findepi** [comment]
@kokosing 
- `Extracted-From: https://github.com/Teradata/presto` would be more appropriate
- first commit has a typo: "Prever"

**Comment by sopel39** [comment]
@kokosing ping me when you consider this ready for review

**Comment by kokosing** [comment]
Travis says:
```
2019-04-09T03:37:04.650-0500	INFO	pool-109-thread-2	io.prestosql.sql.gen.TestExpressionCompiler	FINISHED testBinaryOperatorsIntegerDecimal in 13.76s verified 0 expressions
Terminating due to java.lang.OutOfMemoryError: Java heap space
The 
```

**Comment by electrum** [comment]
The "Minor fixes" commit looks good.

**Comment by electrum** [comment]
Add a test for rolling sum that includes nulls

File: presto-main/src/test/java/io/prestosql/operator/window/TestAggregateWindowFunction.java [hunk 1]
```diff
@@ -57,6 +57,25 @@ public void testCountRowsOrdered()
                         .build());
     }
 
+    @Test
+    public void testCountRowsRolling()
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/test/java/io/prestosql/operator/window/TestAggregateWindowFunction.java]: I also added tests for `PARTITION BY`

**Comment by kokosing** [comment]
@alandpost Thank you! It took us a while to merge your work.

File: presto-main/src/main/java/io/prestosql/operator/aggregation/LongSumAggregation.java [hunk 2]
```diff
@@ -31,27 +32,34 @@
     private LongSumAggregation() {}
 
     @InputFunction
-    public static void sum(@AggregationState NullableLongState state, @SqlType(StandardTypes.BIGINT) long value)
+    public static void sum(@AggregationState LongLongState state, @SqlType(StandardTypes.BIGINT) long value)
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/LongSumAggregation.java:35]: this actually causes significant perf regression as long long stats is much bigger to serialize, cc @martint @dain

Comment (weiatwork) [reply presto-main/src/main/java/io/prestosql/operator/aggregation/LongSumAggregation.java:35]: I agree, and just filed an issue to showcase the problem #18818

**Comment by jlpetz** [comment]
Hi Trino,

This change appears to have resulted in a substantial performance regression for certain queries we have. For example something that was taking 9min is now taking 57min to run, a 6.3x increase.

Is this something that could be looked into?

**Pull request merged**